src/tools/ceph_dedup_tool.cc

```diff
@@ -502,6 +514,564 @@ void ChunkScrub::chunk_scrub_common()
+  }
+
+  return std::make_tuple(objects, next);
+}
+
+std::set<size_t> SampleDedup::sample_object(size_t count) {
+  std::set<size_t> indexes;
+  switch(mode) {
+    // DEEP mode crawls all objects. Crawling has high overhead but finds all
+    // duplication
+    case crawl_mode_t::DEEP: {
+      for (size_t index = 0 ; index < count ; index++) {
+        indexes.insert(index);
+      }
+      break;
+    }
+
+    // SHALLOW mode crawls a few objects. Crawling has low overhead but finds
+    // all less duplication
+    case crawl_mode_t::SHALLOW: {
+      size_t sampling_count = count * sampling_ratio / 100;
+      default_random_engine generator;
+      uniform_int_distribution<size_t> distribution(0, count - 1);
+      while (indexes.size() < sampling_count) {
+        size_t index = distribution(generator);
+        indexes.insert(index);
+      }
+      break;
+    }
+
+    default:
+      assert(false);
+  }
+  return indexes;
+}
+
+void SampleDedup::try_object_dedup_and_accumulate_result(ObjectItem& object) {
+  bufferlist data = read_object(object);
+  if (data.length() == 0) {
+    cerr << __func__ << " skip object " << object.oid
+      << " dedup (read failed)\n";
+    return;
+  }
+  auto chunks = do_cdc(object, data);
+  size_t chunk_total_amount = 0;
+
+  // First, check total size of created chunks
+  for (auto& chunk : chunks) {
+    auto& chunk_data = std::get<0>(chunk);
+    chunk_total_amount += chunk_data.length();
+  }
+  if (chunk_total_amount != data.length()) {
+    cerr << __func__ << " sum of chunked length(" << chunk_total_amount
+      << ") is different from object data length(" << data.length() << ")\n";
+    return;
+  }
+
+  size_t duplicated_size = 0;
+  for (auto& chunk : chunks) {
+    auto& chunk_data = std::get<0>(chunk);
+    std::string fingerprint = generate_fingerprint(chunk_data);
+    std::pair<uint64_t, uint64_t> chunk_boundary = std::get<1>(chunk);
+    chunk_t chunk_info = {
+      .oid = object.oid,
+      .start = chunk_boundary.first,
+      .size = chunk_boundary.second,
+      .fingerprint = fingerprint,
+      .data = chunk_data
+      };
+
+    DEBUG_OUT("check " << chunk_info.oid << " fp " << fingerprint << " "
+      << chunk_info.start << ", " << chunk_info.size << std::endl);
+    if (fp_store.find(fingerprint)) {
+      DEBUG_OUT("duplication oid " << chunk_info.oid <<  " "
+        << chunk_info.fingerprint << " " << chunk_info.start << ", "
+        << chunk_info.size << std::endl);
+      duplicated_size += chunk_data.length();
+    }
+    fp_store.add(chunk_info, duplicable_chunks);
+  }
+
+  size_t object_size = data.length();
+
+  DEBUG_OUT("oid " << object.oid << " object_size " << object_size
+    << " dup size " << duplicated_size << std::endl);
+  // if the chunks in an object are duplicated higher than object_dedup_threshold,
+  // try deduplicate whole object via tier_flush
+  if (check_whole_object_dedupable(duplicated_size, object_size)) {
```

Comment:
I don't understand what purpose this serves.  Why do we need a special mechanism for the case where an object just happens to have a bunch of despicable sections?  Seems to me that this just adds another pathway we'd need to test.

Reply:
Modified this part to clarify the intention. What we intended here is that the dedup tool searches chunks first, then performs either object dedup (if the object is highly redundant) or chunk dedup (some part of the object is redundant). Please see the updated commit (https://github.com/ceph/ceph/pull/45369/commits/9586b61e7351edc101f74cb14147e4b4d414d72e).